discover.py - simplified control flow, added pubsub message for broad…

daniel-montanari opened the pull request:
Basic idea implemented with the dmm drivers.
On loading a driver log the .serial property to the csv.
I haven't given too much thought yet to whether this should be protected, or whether it should just be a method instead. I think a property makes sense though as the serial number shouldn't change once the instrument has been instantiated.

Remaining drivers still need a .serial property added to them so open_visa_instrument in discover.py can use them.

daniel-montanari committed:
discover.py - simplified control flow, added pubsub message for broadcasting the serial number of a visa device.
csv.py - added discover_visa method to log serial numbers
helper.py, fluke_8846a.py - added get_serial methods,

clint-lawrence commented on src/fixate/drivers/dmm/fluke_8846a.py:
```diff
@@ -62,6 +62,7 @@ def __init__(self, instrument, *args, **kwargs):
         self._mode = None
         # Set to True to have explicit error checks on each read
         self.legacy_mode = False
+        self.serial = self.get_serial()
```
I would leave this out. We're not likely to call get_serial often, so if we do need it more than once it's not too big a cost to call out to the instrument multiple times.

clint-lawrence commented on src/fixate/drivers/dmm/fluke_8846a.py:
```diff
@@ -341,3 +342,11 @@ def analog_filter(self, bandwidth=None):
             self._write(self._filters[self.mode] + ":FILT:STAT ON")
         pass
+
+    def get_serial(self):
```
I think we should either parse out the serial number or rename this to get_identity.
Possibly, rename this to get identity, the implement get_serial using this function.

daniel-montanari replied:
I originally was doing some string parsing to get the serial number out, but realised that including everything meant that we could keep the other parameters such as model number, while still letting us filter through them. 
I'll rename it to get_identity

clint-lawrence commented on src/fixate/reporting/csv.py:
```diff
@@ -367,6 +377,7 @@ def register_csv():
     pub.subscribe(writer.reporting.sequence_complete, "Sequence_Complete")
     pub.subscribe(writer.reporting.user_wait_start, "UI_block_start")
     pub.subscribe(writer.reporting.user_wait_end, "UI_block_end")
+    pub.subscribe(writer.reporting.discover_visa, "discover_visa")
```
Rename this message to reflect its function rather than where we're calling it from. "driver_open" maybe?

daniel-montanari replied:
I'll rename it to "driver_open" and change the "VISA" string in the csv writer to "DRIVER"

clint-lawrence commented on src/fixate/drivers/dmm/fluke_8846a.py:
```diff
@@ -341,3 +342,11 @@ def analog_filter(self, bandwidth=None):
+        outg_sw_date is the
+        :return: FLUKE,8846A,S/N
+        """
+        return self.instrument.query("*IDN?")
```
Does query return bytes or string? I'm inclined to think we should be explicit here that get_serial (or get_identity if we implement that) should return a string and not a bytes object, otherwise we potentially need to know about the encoding used elsewhere in the code base.

daniel-montanari replied:
Returns a string

clint-lawrence commented on src/fixate/drivers/dmm/helper.py:
```diff
@@ -67,3 +67,6 @@ def diode(self, low_current=True, high_voltage=False):
     def reset(self):
         raise NotImplementedError
+
+    def get_serial(self):
```
I'm torn on if this should be NotImplementedError or return "". We might not be able to get a serial number from all drivers and for some, it might not make sense. I suppose the argument in favor of keeping it NotImplementedError would be to force the driver implementation to be explicit. It can alwa

daniel-montanari replied:
Some of the other drivers simply pass rather than raising an error.

clint-lawrence reviewed (commented):
To get some test coverage of this, we could use the existing script2log tests. The sticking point is finding a sensible way to inject a test driver. Having a pytest fixture for that might be useful for other purposes also...

daniel-montanari committed:
renamed "discover_visa" pubsub message to "driver_open"
renamed get_serial to get_identity

daniel-montanari committed:
Removed code that should probably be moved to proper tests
Added get_identity methods to other drivers.
Renamed open_visa_instrument to open_instrument as it is also used with PPS

daniel-montanari committed:
Remove code that should be a proper test

daniel-montanari committed:
Black

clint-lawrence commented on src/fixate/drivers/pps/bk_178x.py:
```diff
@@ -355,21 +358,6 @@ def identify(self, as_string=False):
-    mypps.output_ch1 = False
+    def get_identity(self) -> str:
+        # TODO: this won't be formatted how we want
+        return self.identify(as_string=True)
```
Note sure I follow, but I assume you mean to come back to this and fix the formatting?

daniel-montanari replied:
The self.identify function doesn't return a comma delimited string and includes the keys. Will make it consistent later.

clint-lawrence commented on src/fixate/drivers/pps/helper.py:
```diff
@@ -206,3 +198,14 @@ def idn(self):
             )
         )
+
+    def get_identity(self):
```
Implement this by calling idn? or the bind the name as an alias?
self.get_identify = self.idn should work...?

daniel-montanari replied:
PPS was a bit of a mess, bk 178x doesn't seem to even us idn, spd 3303x only mentions it in the api list, which gives me conniptions and I was trying to avoid looking at it. Not how binding names works with inheritance either.

daniel-montanari committed:
Added test for getting the serial numbers using different drivers
Added context manager for lcr as the IDN query requires a longer timeout than the default

daniel-montanari committed:
Add docstring to bk get_identity

daniel-montanari committed:
Remove trailing escape characters
LCR also returns nulls so filter by isprintable

daniel-montanari committed:
Add ignore folder for manual tests that cannot be run automatically
